### ISSUE TITLE
v0.37-experimental: Backport pruning optimizations

### ISSUE THREAD

**Issue by adizere** [cometbft/cometbft#2776]
## What do we want to achieve?

We want to backport the following abstract features:

* pruning + (forced) compaction support
* ensure pruning/compaction is done as a background task, off the critical path, to mitigates the performance impact of this feature

We are not sure if the following feature is worthwhile in improving database access time, so it requires further testing:

* the new "v2" layout

What is the target branch to backport these features? It should be on `v0.37.x-experimental`[^37x].

## Background reading

* The "storage" report[^1] provides context on what we have achieved so far as part of all the storage efficiency improvements over the last 2 quarters.
* The features mentioned below are part of the ADR-101 meta-feature in CometBFT. For completeness, issue [#1420](https://github.com/cometbft/cometbft/issues/1420) shows the comprehensive set of items that would have been needed if we wanted to backport this to `v0.37` and offer it in _non-experimental_ mode.

## Approach

Commit [`e45c01e`](https://github.com/cometbft/cometbft/commit/e45c01e055fc03f6ca5aa20af728d978db457a9d) is probably the best starting point. This was done in PR [#2113](https://github.com/cometbft/cometbft/pull/2113). This commit includes pruning, but not compaction. Let's look for compaction commmits..

You can see `e45c01e` exists on four testing branches, as follows:

1. [`storage/tmp/injective/v0.37.x-testing-newlayout-validator`](https://github.com/cometbft/cometbft/commits/storage/tmp/injective/v0.37.x-testing-newlayout-validator/)
2. [`storage/tmp/injective/v0.37.x-testing-validator`](https://github.com/cometbft/cometbft/commits/storage/tmp/injective/v0.37.x-testing-validator/)
3. [`storage/tmp/v0.37.x-testing-newlayout-validator`](https://github.com/cometbft/cometbft/commits/storage/tmp/v0.37.x-testing-newlayout-validator)
4. [`storage/tmp/v0.37.x-testing-validator`](https://github.com/cometbft/cometbft/commits/storage/tmp/v0.37.x-testing-validator/)

Let's focus on the last two branches, (3) and (4). These are the interesting ones. These two have a common history between commit [`e45c01e`](https://github.com/cometbft/cometbft/commit/e45c01e055fc03f6ca5aa20af728d978db457a9d) and [`793cf23`](https://github.com/cometbft/cometbft/commit/793cf23e811387903a00994576005846b10bacaa). This history of commits is important because it contains compaction

After `793cf23`, the two branches diverge. Probably the rest of commits existing in branch (4) are also of interest, but this needs to be investigated.

If we want to additionally investigate the impact of the new "v2" layout, the associated commits exist on branch (3).

To summarize: we likely need to backport everything between commit [`e45c01e`](https://github.com/cometbft/cometbft/commit/e45c01e055fc03f6ca5aa20af728d978db457a9d) and [`793cf23`](https://github.com/cometbft/cometbft/commit/793cf23e811387903a00994576005846b10bacaa) which exist on branch (4) above. The last three commits (`39c892c`, `9139113`, `1d30feb`) on that branch may also be relevant for backporting.

[^1]: <https://github.com/cometbft/cometbft/tree/main/docs/references/storage>
[^37x]: <https://github.com/cometbft/cometbft/tree/v0.37.x-experimental>

**Comment by adizere** [comment]
Supersedes #1444

**Comment by adizere** [comment]
We originally opened this issue to facilitate and clarify what is needed for backporting pruning optimizations to 0.37 for the Osmosis team. 

After initial investigation (the Osmosis team got to the point of getting everything "working except for tx_index pruning"), the Osmosis team and Comet took the decision it is best not to follow through with the backport and instead focus energy on upgrading Osmosis to Comet 0.38 + SDK 0.50.
